master/internal/hpimportance/hpimportance.go

```diff
@@ -1,15 +1,227 @@
+the importance, read and return the values.
+**/
+
 import (
+	"encoding/csv"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"sort"
+	"strconv"
+
 	"github.com/determined-ai/determined/master/pkg/model"
 )
 
-func computeHPImportance(data []model.HPImportanceTrialData) map[string]float64 {
-	// This is a place-holder for a future Random Forest implementation to actually compute the
-	// importance of each hyperparameter.
-	output := make(map[string]float64)
-	for key := range data[0].Hparams {
-		output[key] = 0.0
+const (
+	maxDiffCompBatch = 4
+	minNumberTrials  = 50
+	idealBatchDiff   = 2
+	nReplications    = 2
+
+	arffFile       = "data.arff"
+	importanceFile = "importance.txt"
+)
+
+// The data needs to be put into an arff format.
+// Arff files require variable declaration at the top,
+// where order does matter. We need to keep define the Hps and
+// keep track of the order so the data columns will match.
+func createDataFile(data map[int][]model.HPImportanceTrialData,
+	experimentConfig *model.ExperimentConfig, workingDir string) (int, error) {
+	f, err := os.Create(path.Join(workingDir, arffFile))
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		err = f.Close()
+		if err != nil {
+			fmt.Printf("failed to close arff file: %s", err.Error())
+		}
+	}()
+
+	// create top of file based on exp config
+	_, err = f.WriteString("@relation data\n\n@attribute metric numeric\n")
+	if err != nil {
+		return 0, err
+	}
+	var hpsOrder []string // HPs must be in the same order for the arff file
+	hps := experimentConfig.Hyperparameters
+
+	for key, element := range hps {
+		var st string
+		switch {
+		case element.ConstHyperparameter != nil:
+			continue
+		case element.CategoricalHyperparameter != nil:
+			hpsOrder = append(hpsOrder, key)
+
+			var values string
+			for _, catVal := range element.CategoricalHyperparameter.Vals {
+				values += fmt.Sprintf("%v,", catVal)
+			}
+			st = fmt.Sprintf("@attribute %s {%s}\n", key, values)
+		default:
+			hpsOrder = append(hpsOrder, key)
+			st = fmt.Sprintf("@attribute %s numeric\n", key)
+		}
+		_, err = f.WriteString(st)
+		if err != nil {
+			return 0, err
+		}
+	}
+	_, err = f.WriteString("@attribute numBatches numeric\n\n")
+	if err != nil {
+		return 0, err
+	}
+
+	// Now we have to add the data to the file
+	_, err = f.WriteString("\n@data\n")
+	if err != nil {
+		return 0, err
+	}
+
+	// First, get batch ids unless better way to get them in go?
+	var batches []int
+	for k := range data {
+		batches = append(batches, k)
+	}
+	sort.Sort(sort.Reverse(sort.IntSlice(batches)))
+
+	totalNumTrials := 0
+	maxNumBatches := batches[0]
+	for _, batchID := range batches {
+		if batchID < maxNumBatches/maxDiffCompBatch {
+			break
+		}
+		if totalNumTrials > minNumberTrials && batchID <= maxNumBatches/idealBatchDiff {
+			break
+		}
+		batchIDStr := fmt.Sprintf("%v", batchID)
+		for _, trial := range data[batchID] {
+			var st string
+			st += fmt.Sprintf("%v", trial.Metric)
```

Comment:
Could use a `strings.Builder` [instead](https://stackoverflow.com/questions/1760757/how-to-efficiently-concatenate-strings-in-go/47798475#47798475). Alternatively, you could wrap the file as a `bufio.NewWriter` if you were concerned about each call to `WriteString` causing a context switch and copy from user space into kernel buffers.

Reply:
I've switched to a string builder everywhere we were doing file IO before. There's still a few Sprintf's where the functionality of the format string adds value, or to minimize the amount of error checking we ought to do. I realize errors from the StringBuilder are unlikely, but as the API is exactly the same as file IO, I replaced the file I/O calls with StringBuilder calls and nothing else.

Reply:
sgtm. it shouldn't be so much data that building the string in memory is an issue, right?

Reply:
Discussed offline: the sprintfs are small. That total data may be larger, but shouldn't exceed other memory usage associated with this feature. But all things considered, bufio is a good idea.